File: code/cvc5mvn/src/main/java/plaid/antlr/FunctionListener.java
package plaid.antlr;

import plaid.PreludeBaseListener;
import plaid.PreludeParser.CommandFuncContext;
import plaid.PreludeParser.ConstraintFuncContext;
import plaid.PreludeParser.ExprFuncContext;
import plaid.PreludeParser.IdentContext;
import plaid.PreludeParser.TypedIdentContext;
import plaid.ast.*;
import plaid.ast.CommandFunction;
import plaid.ast.ConstraintFunction;
import plaid.ast.ExprFunction;
import plaid.ast.Identifier;

import java.util.ArrayList;
import java.util.List;

public class FunctionListener extends PreludeBaseListener {

    private final List<ExprFunction> exprFunctions = new ArrayList<>();
    private final List<CommandFunction> commandFunctions = new ArrayList<>();
    private final List<ConstraintFunction> constraintFunctions = new ArrayList<>();

    public List<ExprFunction> getExprFunctions() {
        return exprFunctions;
    }

    public List<CommandFunction> getCommandFunctions() {
        return commandFunctions;
    }

    public List<ConstraintFunction> getConstraintFunctions() {
        return constraintFunctions;
    }

    private List<Identifier> toIdentifiers(List<IdentContext> contexts) {
        return contexts.stream().skip(1).map(x -> new Identifier(x.getText())).toList();
    }

    private List<TypedIdentifier> toTypedIdentifiers(List<TypedIdentContext> typedIdentContexts){
        return typedIdentContexts.stream().map(x -> new TypedIdentifier(new Identifier(x.ident().getText()), Loader.toType(x.type()))).toList();
    }

    @Override
    public void enterExprFunc(ExprFuncContext ctx) {
        exprFunctions.add(new ExprFunction(
                new Identifier(ctx.ident(0).getText()), // fname
                toIdentifiers(ctx.ident()), // parameters
                Loader.toExpression(ctx.expr())));
    }

    @Override
    public void enterCommandFunc(CommandFuncContext ctx) {
        // command function may not have the precondition and postcondition
        // then interpret null constraint as T
       Constraint precondition = ctx.precondsection() == null? null : Loader.toConstraintExpression(ctx.precondsection().constraintExpr());
       Constraint postcondition = ctx.postcondsection() == null? null : Loader.toConstraintExpression(ctx.postcondsection().constraintExpr());     
        
        
        commandFunctions.add(new CommandFunction(
                new Identifier(ctx.ident().getText()), // fname
                toTypedIdentifiers(ctx.typedIdent()),
                Loader.toCommand(ctx.command()),
                precondition,
                postcondition));
    }

    @Override
    public void enterConstraintFunc(ConstraintFuncContext ctx) {
        constraintFunctions.add(new ConstraintFunction(
                new Identifier(ctx.ident(0).getText()),
                toIdentifiers(ctx.ident()),
                Loader.toConstraintExpression(ctx.constraintExpr())));
    }


}

File: code/cvc5mvn/src/main/scala/plaid/antlr/FunctionListener.scala
package plaid.antlr

import plaid.PreludeBaseListener
import plaid.PreludeParser.{CommandFuncContext, ConstraintFuncContext, ExprFuncContext, IdentContext, TypedIdentContext}
import plaid.ast.*

import scala.jdk.CollectionConverters.*

class FunctionListener extends PreludeBaseListener {

  val exprFunctions = scala.collection.mutable.ListBuffer[ExprFunction]()
  val commandFunctions = scala.collection.mutable.ListBuffer[CommandFunction]()
  val constraintFunctions = scala.collection.mutable.ListBuffer[ConstraintFunction]()

  private def toIdentifiers(contexts: java.util.List[IdentContext]): List[Identifier] =
    contexts.asScala.drop(1).map(ctx => Identifier(ctx.getText)).toList

  private def toTypedIdentifiers(contexts: java.util.List[TypedIdentContext]): List[TypedIdentifier] =
    contexts.asScala.map { ctx =>
      TypedIdentifier(Identifier(ctx.ident().getText), Load.typeMarker(ctx.`type`()))
    }.toList

  override def enterExprFunc(ctx: ExprFuncContext): Unit = {
    exprFunctions += ExprFunction(
      Identifier(ctx.ident(0).getText), // function name
      toIdentifiers(ctx.ident()),           // parameters
      Load.expression(ctx.expr())       // body
    )
  }

  override def enterCommandFunc(ctx: CommandFuncContext): Unit = {
    val precondition =
      Option(ctx.precondsection()).map(_.constraintExpr()).map(Load.constraint).orNull
    val postcondition =
      Option(ctx.postcondsection()).map(_.constraintExpr()).map(Load.constraint).orNull

    commandFunctions += CommandFunction(
      Identifier(ctx.ident().getText),
      toTypedIdentifiers(ctx.typedIdent()),
      Load.command(ctx.command()),
      precondition,
      postcondition
    )
  }

  override def enterConstraintFunc(ctx: ConstraintFuncContext): Unit = {
    constraintFunctions += ConstraintFunction(
      Identifier(ctx.ident(0).getText),
      toIdentifiers(ctx.ident()),
      Load.constraint(ctx.constraintExpr())
    )
  }
}
